// @chainsafe/bls/lib/blst/signature.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const blst = __importStar(require("@chainsafe/blst"));
const helpers_1 = require("../helpers");
const publicKey_1 = require("./publicKey");
const errors_1 = require("../errors");
class Signature {
    constructor(value) {
        this.affine = value;
    }
    static fromBytes(bytes) {
        return new Signature(blst.Signature.fromBytes(bytes));
    }
    static fromHex(hex) {
        return this.fromBytes(helpers_1.hexToBytes(hex));
    }
    static aggregate(signatures) {
        if (signatures.length === 0) {
            throw new errors_1.EmptyAggregateError();
        }
        const agg = blst.AggregateSignature.fromSignatures(signatures.map((sig) => sig.affine));
        return new Signature(agg.toSignature());
    }
    static verifyMultipleSignatures(publicKeys, messages, signatures) {
        return blst.verifyMultipleAggregateSignatures(messages, publicKeys.map((publicKey) => publicKey.affine), signatures.map((signature) => signature.affine));
    }
    verify(publicKey, message) {
        // Individual infinity signatures are NOT okay. Aggregated signatures MAY be infinity
        if (this.affine.value.is_inf()) {
            throw new errors_1.ZeroSignatureError();
        }
        return this.aggregateVerify([message], [publicKey.affine]);
    }
    verifyAggregate(publicKeys, message) {
        const agg = publicKey_1.PublicKey.aggregate(publicKeys);
        return this.aggregateVerify([message], [agg.affine]);
    }
    verifyMultiple(publicKeys, messages) {
        return this.aggregateVerify(messages, publicKeys.map((pk) => pk.affine));
    }
    toBytes() {
        return this.affine.toBytes();
    }
    toHex() {
        return helpers_1.bytesToHex(this.toBytes());
    }
    aggregateVerify(msgs, pks) {
        // If this set is simply an infinity signature and infinity publicKey then skip verification.
        // This has the effect of always declaring that this sig/publicKey combination is valid.
        // for Eth2.0 specs tests
        if (this.affine.value.is_inf() && pks.length === 1 && pks[0].value.is_inf()) {
            return true;
        }
        return blst.aggregateVerify(msgs, pks, this.affine);
    }
}
exports.Signature = Signature;
