// @myetherwallet/eth2-keystore/dist/generateKeystore.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _bls = require("@chainsafe/bls");

var _crypto = require("crypto");

var _scryptJs = require("scrypt-js");

var _ethereumjsUtil = require("ethereumjs-util");

var _uuid = require("uuid");

var _utils = require("./utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var generate = /*#__PURE__*/function () {
  var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(privateKey, password) {
    var params,
        path,
        v4Defaults,
        derivedKey,
        kdfParams,
        cipher,
        ciphertext,
        _args = arguments;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            params = _args.length > 2 && _args[2] !== undefined ? _args[2] : {};
            path = _args.length > 3 && _args[3] !== undefined ? _args[3] : '';
            password = (0, _utils.normalizePassword)(password);
            v4Defaults = {
              cipher: 'aes-128-ctr',
              kdf: 'scrypt',
              salt: (0, _crypto.randomBytes)(32),
              iv: (0, _crypto.randomBytes)(16),
              dklen: 32,
              c: 262144,
              n: 262144,
              r: 8,
              p: 1,
              uuid: (0, _crypto.randomBytes)(16)
            };
            Object.assign(v4Defaults, params);

            if (!(v4Defaults.kdf === _utils.KDFFunctions.PBKDF)) {
              _context.next = 10;
              break;
            }

            kdfParams = (0, _utils.kdfParamsForPBKDF)(v4Defaults);
            derivedKey = (0, _crypto.pbkdf2Sync)(Buffer.from(password), kdfParams.salt, kdfParams.c, kdfParams.dklen, 'sha256');
            _context.next = 15;
            break;

          case 10:
            kdfParams = (0, _utils.kdfParamsForScrypt)(v4Defaults);
            _context.next = 13;
            return (0, _scryptJs.scrypt)(Buffer.from(password), kdfParams.salt, kdfParams.n, kdfParams.r, kdfParams.p, kdfParams.dklen);

          case 13:
            derivedKey = _context.sent;
            derivedKey = Buffer.from(derivedKey);

          case 15:
            cipher = (0, _crypto.createCipheriv)(v4Defaults.cipher, derivedKey.slice(0, 16), v4Defaults.iv);
            ciphertext = (0, _utils.runCipherBuffer)(cipher, privateKey);
            _context.next = 19;
            return (0, _bls.init)('herumi');

          case 19:
            return _context.abrupt("return", {
              crypto: {
                kdf: {
                  "function": v4Defaults.kdf,
                  params: _objectSpread(_objectSpread({}, kdfParams), {}, {
                    salt: kdfParams.salt.toString('hex')
                  }),
                  message: ''
                },
                checksum: {
                  "function": 'sha256',
                  params: {},
                  message: (0, _ethereumjsUtil.sha256)(Buffer.concat([derivedKey.slice(16), ciphertext])).toString('hex')
                },
                cipher: {
                  "function": v4Defaults.cipher,
                  params: {
                    iv: v4Defaults.iv.toString('hex')
                  },
                  message: ciphertext.toString('hex')
                }
              },
              description: '',
              pubkey: Buffer.from(_bls.SecretKey.fromBytes(privateKey).toPublicKey().toBytes()).toString('hex'),
              path: path,
              uuid: (0, _uuid.v4)({
                random: v4Defaults.uuid
              }),
              version: 4
            });

          case 20:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));

  return function generate(_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();

var _default = generate;
exports["default"] = _default;
