// @myetherwallet/eth2-keystore/dist/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "generateKeystore", {
  enumerable: true,
  get: function get() {
    return _generateKeystore["default"];
  }
});
Object.defineProperty(exports, "verifyKeystore", {
  enumerable: true,
  get: function get() {
    return _verifyKeystore["default"];
  }
});
exports["default"] = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var bip39 = _interopRequireWildcard(require("bip39"));

var _blsHdKey = require("@chainsafe/bls-hd-key");

var _bls = require("@chainsafe/bls");

var _generateKeystore = _interopRequireDefault(require("./generateKeystore"));

var _verifyKeystore = _interopRequireDefault(require("./verifyKeystore"));

var Keystore = /*#__PURE__*/function () {
  function Keystore() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        _ref$mnemonic = _ref.mnemonic,
        mnemonic = _ref$mnemonic === void 0 ? '' : _ref$mnemonic,
        _ref$password = _ref.password,
        password = _ref$password === void 0 ? '' : _ref$password,
        _ref$bits = _ref.bits,
        bits = _ref$bits === void 0 ? 256 : _ref$bits,
        _ref$lang = _ref.lang,
        lang = _ref$lang === void 0 ? 'english' : _ref$lang;

    (0, _classCallCheck2["default"])(this, Keystore);
    bip39.setDefaultWordlist(lang);

    if (mnemonic === '') {
      mnemonic = bip39.generateMnemonic(bits);
    }

    this.mnemonic = mnemonic;
    this.bits = bits;
    this.seed = bip39.mnemonicToSeedSync(mnemonic, password);
    this.masterKey = (0, _blsHdKey.deriveMasterSK)(this.seed);
  }

  (0, _createClass2["default"])(Keystore, [{
    key: "getEntropy",
    value: function () {
      var _getEntropy = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                return _context.abrupt("return", bip39.mnemonicToEntropy(this.mnemonic));

              case 1:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function getEntropy() {
        return _getEntropy.apply(this, arguments);
      }

      return getEntropy;
    }()
  }, {
    key: "getSeed",
    value: function () {
      var _getSeed = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                return _context2.abrupt("return", this.seed);

              case 1:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function getSeed() {
        return _getSeed.apply(this, arguments);
      }

      return getSeed;
    }()
  }, {
    key: "getMnemonic",
    value: function () {
      var _getMnemonic = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                return _context3.abrupt("return", this.mnemonic);

              case 1:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function getMnemonic() {
        return _getMnemonic.apply(this, arguments);
      }

      return getMnemonic;
    }()
  }, {
    key: "getPath",
    value: function () {
      var _getPath = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var idx,
            isSigning,
            withdrawalPath,
            _args4 = arguments;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                idx = _args4.length > 0 && _args4[0] !== undefined ? _args4[0] : 0;
                isSigning = _args4.length > 1 && _args4[1] !== undefined ? _args4[1] : true;
                withdrawalPath = "m/12381/3600/".concat(idx, "/0");
                return _context4.abrupt("return", isSigning ? withdrawalPath + '/0' : withdrawalPath);

              case 4:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4);
      }));

      function getPath() {
        return _getPath.apply(this, arguments);
      }

      return getPath;
    }()
  }, {
    key: "getChildKey",
    value: function () {
      var _getChildKey = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5() {
        var idx,
            isSigning,
            path,
            _args5 = arguments;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                idx = _args5.length > 0 && _args5[0] !== undefined ? _args5[0] : 0;
                isSigning = _args5.length > 1 && _args5[1] !== undefined ? _args5[1] : true;
                _context5.next = 4;
                return this.getPath(idx, isSigning);

              case 4:
                path = _context5.sent;
                return _context5.abrupt("return", (0, _blsHdKey.deriveChildSKMultiple)(this.masterKey, (0, _blsHdKey.pathToIndices)(path)));

              case 6:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this);
      }));

      function getChildKey() {
        return _getChildKey.apply(this, arguments);
      }

      return getChildKey;
    }()
  }, {
    key: "getPublicKey",
    value: function () {
      var _getPublicKey = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6() {
        var idx,
            isSigning,
            _args6 = arguments;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                idx = _args6.length > 0 && _args6[0] !== undefined ? _args6[0] : 0;
                isSigning = _args6.length > 1 && _args6[1] !== undefined ? _args6[1] : true;
                _context6.next = 4;
                return (0, _bls.init)('herumi');

              case 4:
                _context6.t0 = Buffer;
                _context6.t1 = _bls.SecretKey;
                _context6.next = 8;
                return this.getChildKey(idx, isSigning);

              case 8:
                _context6.t2 = _context6.sent;
                _context6.t3 = _context6.t1.fromBytes.call(_context6.t1, _context6.t2).toPublicKey().toBytes();
                return _context6.abrupt("return", _context6.t0.from.call(_context6.t0, _context6.t3));

              case 11:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this);
      }));

      function getPublicKey() {
        return _getPublicKey.apply(this, arguments);
      }

      return getPublicKey;
    }()
  }, {
    key: "toSigningKeystore",
    value: function () {
      var _toSigningKeystore = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee7(password) {
        var idx,
            params,
            childKey,
            path,
            _args7 = arguments;
        return _regenerator["default"].wrap(function _callee7$(_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                idx = _args7.length > 1 && _args7[1] !== undefined ? _args7[1] : 0;
                params = _args7.length > 2 && _args7[2] !== undefined ? _args7[2] : {};
                _context7.next = 4;
                return this.getChildKey(idx, true);

              case 4:
                childKey = _context7.sent;
                _context7.next = 7;
                return this.getPath(idx, true);

              case 7:
                path = _context7.sent;
                return _context7.abrupt("return", (0, _generateKeystore["default"])(childKey, password, params, path));

              case 9:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee7, this);
      }));

      function toSigningKeystore(_x) {
        return _toSigningKeystore.apply(this, arguments);
      }

      return toSigningKeystore;
    }()
  }, {
    key: "toWithdrawalKeystore",
    value: function () {
      var _toWithdrawalKeystore = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee8(password) {
        var idx,
            params,
            childKey,
            path,
            _args8 = arguments;
        return _regenerator["default"].wrap(function _callee8$(_context8) {
          while (1) {
            switch (_context8.prev = _context8.next) {
              case 0:
                idx = _args8.length > 1 && _args8[1] !== undefined ? _args8[1] : 0;
                params = _args8.length > 2 && _args8[2] !== undefined ? _args8[2] : {};
                _context8.next = 4;
                return this.getChildKey(idx, false);

              case 4:
                childKey = _context8.sent;
                _context8.next = 7;
                return this.getPath(idx, false);

              case 7:
                path = _context8.sent;
                return _context8.abrupt("return", (0, _generateKeystore["default"])(childKey, password, params, path));

              case 9:
              case "end":
                return _context8.stop();
            }
          }
        }, _callee8, this);
      }));

      function toWithdrawalKeystore(_x2) {
        return _toWithdrawalKeystore.apply(this, arguments);
      }

      return toWithdrawalKeystore;
    }()
  }]);
  return Keystore;
}();

var _default = Keystore;
exports["default"] = _default;
