// @chainsafe/bls/lib/herumi/publicKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const context_1 = require("./context");
const constants_1 = require("../constants");
const helpers_1 = require("../helpers");
const errors_1 = require("../errors");
class PublicKey {
    constructor(value) {
        if (value.isZero()) {
            throw new errors_1.ZeroPublicKeyError();
        }
        this.value = value;
    }
    static fromBytes(bytes) {
        if (bytes.length !== constants_1.PUBLIC_KEY_LENGTH) {
            throw new errors_1.InvalidLengthError("PublicKey", constants_1.PUBLIC_KEY_LENGTH);
        }
        const context = context_1.getContext();
        const publicKey = new context.PublicKey();
        if (!helpers_1.isZeroUint8Array(bytes)) {
            publicKey.deserialize(bytes);
        }
        return new PublicKey(publicKey);
    }
    static fromHex(hex) {
        return this.fromBytes(helpers_1.hexToBytes(hex));
    }
    static aggregate(publicKeys) {
        if (publicKeys.length === 0) {
            throw new errors_1.EmptyAggregateError();
        }
        const agg = new PublicKey(publicKeys[0].value.clone());
        for (const pk of publicKeys.slice(1)) {
            agg.value.add(pk.value);
        }
        return agg;
    }
    toBytes() {
        return this.value.serialize();
    }
    toHex() {
        return helpers_1.bytesToHex(this.toBytes());
    }
}
exports.PublicKey = PublicKey;
