// @chainsafe/bls/lib/herumi/signature.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("../constants");
const bls_eth_wasm_1 = require("bls-eth-wasm");
const context_1 = require("./context");
const helpers_1 = require("../helpers");
const errors_1 = require("../errors");
class Signature {
    constructor(value) {
        if (!value.isValidOrder()) {
            throw new errors_1.InvalidOrderError();
        }
        this.value = value;
    }
    static fromBytes(bytes) {
        if (bytes.length !== constants_1.SIGNATURE_LENGTH) {
            throw new errors_1.InvalidLengthError("Signature", constants_1.SIGNATURE_LENGTH);
        }
        const context = context_1.getContext();
        const signature = new context.Signature();
        if (!helpers_1.isZeroUint8Array(bytes)) {
            signature.deserialize(bytes);
        }
        return new Signature(signature);
    }
    static fromHex(hex) {
        return this.fromBytes(helpers_1.hexToBytes(hex));
    }
    static aggregate(signatures) {
        if (signatures.length === 0) {
            throw new errors_1.EmptyAggregateError();
        }
        const context = context_1.getContext();
        const signature = new context.Signature();
        signature.aggregate(signatures.map((sig) => sig.value));
        return new Signature(signature);
    }
    static verifyMultipleSignatures(publicKeys, messages, signatures) {
        return bls_eth_wasm_1.multiVerify(publicKeys.map((publicKey) => publicKey.value), signatures.map((signature) => signature.value), messages);
    }
    verify(publicKey, message) {
        return publicKey.value.verify(this.value, message);
    }
    verifyAggregate(publicKeys, message) {
        return this.value.fastAggregateVerify(publicKeys.map((key) => key.value), message);
    }
    verifyMultiple(publicKeys, messages) {
        return this.value.aggregateVerifyNoCheck(publicKeys.map((key) => key.value), helpers_1.concatUint8Arrays(messages));
    }
    toBytes() {
        return this.value.serialize();
    }
    toHex() {
        return helpers_1.bytesToHex(this.toBytes());
    }
}
exports.Signature = Signature;
